### PR TITLE
ipnlocal: store last suggested exit node in local backend

### PR DESCRIPTION
**Pull request by clairew**
tailscale/corp#19681

Start storing last suggested exit node id, and fallback on the suggestion if the last exit node suggestion exists.

### CODE REVIEW

File: ipn/ipnlocal/local.go [hunk 1]
```diff
@@ -324,6 +330,10 @@ type LocalBackend struct {
 
 	// outgoingFiles keeps track of Taildrop outgoing files keyed to their OutgoingFile.ID
 	outgoingFiles map[string]*ipn.OutgoingFile
+
+	// lastSuggestedExitNode stores the last suggested exit node ID and name.
+	// lastSuggestedExitNode updates whenever the suggestion changes.
+	lastSuggestedExitNode lastSuggestedExitNode
```
Comment (agottardo) [inline ipn/ipnlocal/local.go:336]: Do we have to reset this value when switching between user profiles? (different tailnets won't have the same exit nodes)

Comment (clairew) [reply ipn/ipnlocal/local.go:336]: decided on clearing the last suggested exit node everytime there's a profile switch

**Review by agottardo — APPROVED** [review]

File: ipn/ipnlocal/local.go [hunk 2]
```diff
@@ -6347,13 +6359,41 @@ func (b *LocalBackend) SuggestExitNode() (response apitype.ExitNodeSuggestionRes
...
+	b.mu.Unlock()
+	return res, err
+}
+
+// suggestLastExitNode formats a response with the last suggested exit node's ID and name.
+// Used as a fallback before returning a nil response and error.
+func suggestLastExitNode(lastSuggestedExitNode lastSuggestedExitNode) (res apitype.ExitNodeSuggestionResponse, err error) {
+	if lastSuggestedExitNode.id != "" && lastSuggestedExitNode.name != "" {
+		res.ID = lastSuggestedExitNode.id
+		res.Name = lastSuggestedExitNode.name
```
Comment (bradfitz) [inline ipn/ipnlocal/local.go:6393]: This looks wrong. Aren't these supposed to be guarded by the mutex? These are being read without mu held, aren't they?

If so, the method should also be named with the "Locked" suffix.

File: wgengine/magicsock/magicsock.go [hunk 3]
```diff
@@ -3088,3 +3088,9 @@ func (c *Conn) GetLastNetcheckReport(ctx context.Context) *netcheck.Report {
 	}
 	return lastReport
 }
+
+// SetLastNetcheckReport sets local backend's last netcheck report.
+// Used for testing purposes.
+func (c *Conn) SetLastNetcheckReport(ctx context.Context, report netcheck.Report) {
+	c.lastNetCheckReport.Store(&report)
```
Comment (bradfitz) [inline wgengine/magicsock/magicsock.go:3095]: this should be named "ForTest" at the end, and should take a pointer instead, and document that it takes ownership of the report.

File: ipn/ipnlocal/local.go (hunk 2 above)
Comment (clairew) [reply ipn/ipnlocal/local.go:6393]: this function just formats the values from lastSuggestedExitNode into the apitype.ExitNodeSuggestionResponse. This is called after line 6362, where the mutex is held and is reading local backend's last suggested exit node value.

File: wgengine/magicsock/magicsock.go (hunk 3 above)
Comment (clairew) [reply wgengine/magicsock/magicsock.go:3095]: going to update naming in another pr

**Pull request merged**